controller/program/src/instructions/liquidity/add.rs

```diff
@@ -0,0 +1,368 @@
+    )
+    .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidReserves))?;
+
+    let expected_pks = NewAddLiquidityIxPreAccsBuilder::start()
+        // These can be arbitrary accs bc they belong to the user adding liquidity to INF
+        .with_signer(abr.get(*ix_prefix.signer()).key())
+        .with_lst_mint(&lst_state.mint)
+        .with_lst_acc(abr.get(*ix_prefix.lst_acc()).key())
+        .with_lp_acc(abr.get(*ix_prefix.lp_acc()).key())
+        .with_lp_token_mint(&pool.lp_token_mint)
+        .with_protocol_fee_accumulator(&expected_protocol_fee_accumulator)
+        .with_lst_token_program(&TOKEN_PROGRAM)
+        .with_lp_token_program(&TOKEN_PROGRAM)
+        .with_pool_state(&POOL_STATE_ID)
+        .with_lst_state_list(&LST_STATE_LIST_ID)
+        .with_pool_reserves(&expected_reserves)
+        .build();
+
+    verify_pks(abr, &ix_prefix.0, &expected_pks.0)?;
+
+    let (lst_cal_all, pricing_all) = suf
+        // Adding +1 here bc we need to take into account the program as well
+        .split_at_checked((ix_args.lst_value_calc_accs + 1).into())
+        .ok_or(NOT_ENOUGH_ACCOUNT_KEYS)?;
+
+    let (lst_calc_prog, lst_calc_acc) = lst_cal_all.split_first().ok_or(NOT_ENOUGH_ACCOUNT_KEYS)?;
+
+    let (pricing_prog, pricing_accs) = pricing_all.split_first().ok_or(NOT_ENOUGH_ACCOUNT_KEYS)?;
+
+    verify_pks(
+        abr,
+        &[*lst_calc_prog, *pricing_prog],
+        &[&lst_state.sol_value_calculator, &pool.pricing_program],
+    )?;
+
+    verify_not_rebalancing_and_not_disabled(pool)?;
+    verify_not_input_disabled(lst_state)?;
+
+    Ok(AddLiquidityIxAccounts {
+        ix_prefix,
+        lst_calc_prog: *lst_calc_prog,
+        lst_calc: lst_calc_acc,
+        pricing_prog: *pricing_prog,
+        pricing: pricing_accs,
+    })
+}
+
+#[inline]
+#[allow(deprecated)]
+pub fn process_add_liquidity(
+    abr: &mut Abr,
+    accounts: &[AccountHandle],
+    ix_args: AddLiquidityIxArgs,
+    cpi: &mut Cpi,
+) -> Result<(), ProgramError> {
+    let AddLiquidityIxAccounts {
+        ix_prefix,
+        lst_calc_prog,
+        lst_calc,
+        pricing_prog,
+        pricing,
+    } = add_liquidity_accs_checked(abr, accounts, ix_args)?;
+
+    let lp_lst_supply = RawTokenAccount::of_acc_data(abr.get(*ix_prefix.pool_reserves()).data())
+        .and_then(TokenAccount::try_from_raw)
+        .map(|a| a.amount())
+        .ok_or(INVALID_ACCOUNT_DATA)?;
+
+    lst_sync_sol_val_unchecked(
+        abr,
+        cpi,
+        SyncSolValueIxAccs {
+            ix_prefix: NewSyncSolValueIxPreAccsBuilder::start()
+                .with_lst_mint(*ix_prefix.lst_mint())
+                .with_pool_state(*ix_prefix.pool_state())
+                .with_lst_state_list(*ix_prefix.lst_state_list())
+                .with_pool_reserves(*ix_prefix.pool_reserves())
+                .build(),
+            calc_prog: lst_calc_prog,
+            calc: lst_calc,
+        },
+        ix_args.lst_index as usize,
+    )?;
```

Comment:
you can save the built `SyncSolValueIxPreAccs` + `SyncSolValueIxAccs` for the later sync sol value